pkg/buses/natss/bus.go

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2018 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package natss
+
+import (
+	"time"
+	"go.uber.org/zap"
+
+	stan "github.com/nats-io/go-nats-streaming"
+	stanutil "github.com/knative/eventing/pkg/buses/natss/stanutil"
+	"github.com/knative/eventing/pkg/buses"
+)
+
+// BusType is the type of the stub bus
+const BusType = "natss"
+
+type NatssBus struct {
+	natsConn *stan.Conn
+	subscribers map[string]*stan.Subscription
+
+	ref         buses.BusReference
+	dispatcher  buses.BusDispatcher
+	provisioner buses.BusProvisioner
+
+	logger *zap.SugaredLogger
+}
+
+var (
+	natsConn *stan.Conn
+)
+
+func NewNatssBusProvisioner(ref buses.BusReference, opts *buses.BusOpts) (*NatssBus, error) {
+	bus := &NatssBus{
+		ref: ref,
+		subscribers: make(map[string]*stan.Subscription),
+	}
+	eventHandlers := buses.EventHandlerFuncs{
+		ProvisionFunc: func(channel buses.ChannelReference, parameters buses.ResolvedParameters) error {
+			bus.logger.Infof("Provision channel %q\n", channel.Name)
+			bus.logger.Info("channel=%+v; parameters=%+v", channel, parameters)
+			// TODO create a NATSS subject
+			return nil
+		},
+		UnprovisionFunc: func(channel buses.ChannelReference) error {
+			bus.logger.Infof("Unprovision channel %q\n", channel.Name)
+			bus.logger.Info("channel=%+v", channel)
+			// TODO delete a NATSS subject
+			return nil
+		},
+		SubscribeFunc: func(channel buses.ChannelReference, subscription buses.SubscriptionReference, parameters buses.ResolvedParameters) error {
```

Comment:
I think SubscribeFunc and UnsubscribeFunc should be set in the Dispatcher, not the Provisioner.

Reply:
Moved